llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,245 @@
+//===-- LoopIterCountAssumptions.cpp - add Loop assumptions -----*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass converts Loop Iteration Count Metadata to Assumptions which can be
+// picked up by Loop Rotate to remove Loop Guards.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+namespace {
+
+struct BundledSCEV {
+  /// AddRec Scalar Evolution Expression of the IV
+  const SCEVAddRecExpr *AddRecExpr = nullptr;
+  /// Operand Index of the IV in the compare Instruction
+  int IVOperandIndex = -1;
+
+  BundledSCEV(const SCEVAddRecExpr *AddRecExpr, int IVOperandIndex)
+      : AddRecExpr(AddRecExpr), IVOperandIndex(IVOperandIndex) {}
+};
+
+std::string getFunctionAndBlockNames(BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// Get AddRec Scalar Evolution Expression, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution &SE) {
+  if (!SE.isSCEVable(Op->getType()))
```

Comment:
What happens if one calls `getSCEV` on a non "SCEVable" value? Could this function just be replaced by `dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(Op))`?

Reply:
that is convinient